### ISSUE TITLE
Implementing Knowledge Distillation in YOLOv11: Training the Student Model with Teacher Predictions

### ISSUE THREAD

**Issue by aziziselma** [ultralytics/ultralytics#19386]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Bug

Hello everyone,  

I wanted to apply Knowledge Distillation (KD) to the YOLOv11n model using a YOLOv11x model as the teacher. However, when I try to train the student model using the detection trainer, the student model is always trained without KD.  

Could you please guide me on how to modify the training function so that the student model is trained using both its own predictions and the teacher model's predictions?  

Thank you in advance!

### Environment

Ultralytics 8.3.78 🚀 Python-3.11.11 torch-2.5.1+cu124 CUDA:0 (Tesla T4, 15095MiB)
Setup complete ✅ (2 CPUs, 12.7 GB RAM, 33.6/112.6 GB disk)

OS                  Linux-6.1.85+-x86_64-with-glibc2.35
Environment         Colab
Python              3.11.11
Install             pip
RAM                 12.67 GB
Disk                33.6/112.6 GB
CPU                 Intel Xeon 2.20GHz
CPU count           2
GPU                 Tesla T4, 15095MiB
GPU count           1
CUDA                12.4

numpy               ✅ 1.26.4<=2.1.1,>=1.23.0
matplotlib          ✅ 3.10.0>=3.3.0
opencv-python       ✅ 4.11.0.86>=4.6.0
pillow              ✅ 11.1.0>=7.1.2
pyyaml              ✅ 6.0.2>=5.3.1
requests            ✅ 2.32.3>=2.23.0
scipy               ✅ 1.13.1>=1.4.1
torch               ✅ 2.5.1+cu124>=1.8.0
torch               ✅ 2.5.1+cu124!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.20.1+cu124>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 5.9.5
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.2>=1.1.4
seaborn             ✅ 0.13.2>=0.11.0
ultralytics-thop    ✅ 2.0.14>=2.0.0

### Minimal Reproducible Example

class KnowledgeDistillationTrainer(DetectionTrainer):
    def __init__(self, cfg, model, optimizer, train_loader, teacher_model, fgdloss):
        super().__init__(cfg=cfg)
        self.model = model
        self.optimizer = optimizer
        self.train_loader = train_loader
        self.teacher_model = teacher_model
        self.fgdloss = fgdloss

    def train(self):
        for epoch in range(self.args.epochs):
            self.model.train()  # Mettre le modèle étudiant en mode entraînement

            running_loss = 0.0
            for batch in self.train_loader:
                images = batch["img"].to(self.device)
                targets = batch["targets"].to(self.device)

                # Prédictions du modèle étudiant et enseignant
                student_preds = self.model(images)
                teacher_preds = self.teacher_model(images).detach()

                # Calcul de la perte de distillation
                loss = self.fgdloss(student_preds, teacher_preds, batch)
                self.optimizer.zero_grad()
                loss.backward()
                self.optimizer.step()

                running_loss += loss.item()

            LOGGER.info(f"Epoch {epoch + 1}/{self.args.epochs}, Loss: {running_loss / len(self.train_loader)}")

      
            if (epoch + 1) % 10 == 0:
                torch.save(self.model.state_dict(), f"student_model_epoch_{epoch + 1}.pth")
trainer = KnowledgeDistillationTrainer(
    cfg=cfg,
    model=student_model, 
    optimizer=optimizer, 
    train_loader=train_loader, 
    teacher_model=teacher_model, 
    fgdloss=fgdloss
)

trainer.train()

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @aziziselma, thank you for your interest in Ultralytics 🚀! We appreciate you taking the time to share your Knowledge Distillation implementation inquiry. This is an automated response to help guide you while an Ultralytics engineer reviews your request.

To ensure we can assist you effectively, please make sure you've included a [minimum reproducible example (MRE)](https://docs.ultralytics.com/help/minimum-reproducible-example/) demonstrating the issue you're encountering with YOLOv11. Your provided code snippet for the trainer is great! However, if possible, share simplified runnable code so we can better assist.

In the meantime, here are a few resources that may be helpful for your request:

- Visit our [Docs](https://docs.ultralytics.com/) for general usage and training help, including [model training tips](https://docs.ultralytics.com/guides/model-training-tips/).
- For advanced customization discussions such as your Knowledge Distillation setup, check out or share your insights with the Ultralytics community:
  - [Discord](https://discord.com/invite/ultralytics) 🎧 for real-time chat
  - [Discourse](https://community.ultralytics.com/) for in-depth discussions
  - [Subreddit](https://reddit.com/r/Ultralytics) to connect with fellow users

## Upgrade

Before proceeding, ensure your environment is updated to the latest Ultralytics package and dependencies:

```bash
pip install -U ultralytics
```

This will ensure the latest features and fixes are applied to your workflow.

## Verified Environments

You mentioned using Colab, which is great! YOLO also supports many verified environments if needed:

- **Notebooks**: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Docker** and other setups: See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>  

This badge reflects the status of our Continuous Integration (CI) pipeline, ensuring YOLO runs across multiple modes, tasks, and systems every 24 hours.

We look forward to collaborating with you and hope an engineer can assist you further shortly. Thank you again for your contribution to pushing YOLO's boundaries! 💪

**Comment by glenn-jocher** [comment]
@aziziselma to implement knowledge distillation in YOLO11, you'll need to modify the loss function to incorporate teacher outputs. Here's a concise approach:

1. **Override the loss computation**: Create a custom `DetectionTrainer` subclass that integrates teacher model predictions during loss calculation. The Ultralytics engine allows customization through trainer and loss function modifications.

2. **Modify forward pass**: Access both student and teacher outputs in your training loop. We recommend implementing this through a callback or custom loss function that leverages the existing framework shown in the DetectionTrainer documentation.

For implementation specifics, refer to our model training documentation and consider contributing your working solution via PR once validated. The community would appreciate this enhancement!

[Learn more about customizing YOLO training](https://docs.ultralytics.com/modes/train/) and [contributing guidelines](https://docs.ultralytics.com/help/contributing/).

**Comment by 7rkMnpl** [comment]
@aziziselma , thank you for post..  i am also looking something related to KD ,  wanted to know are you able to rectify the issue..

**Comment by Youho99** [comment]
Look at this repo : https://github.com/danielsyahputra/yolo-distiller

**Comment by glenn-jocher** [comment]
@7rkMnpl @aziziselma Knowledge Distillation is a great approach for creating efficient YOLO models. To implement KD properly within the Ultralytics framework, you need to focus on a few key components:

You'll want to subclass `DetectionTrainer` and override the `_do_train_epoch` method to add distillation logic. The key is to run teacher inference before calculating student loss, then incorporate both into your training.

The repo @Youho99 shared provides a good starting point, but for YOLO11 you'll need to integrate this with our current architecture. Instead of creating a completely separate trainer class, try extending our existing framework:

```python
class KDDetectionTrainer(DetectionTrainer):
    def __init__(self, cfg=DEFAULT_CFG, overrides=None):
        super().__init__(cfg, overrides)
        # Load your teacher model here
        self.teacher = YOLO('yolo11x.pt')
        self.teacher.model.eval()
        
    def preprocess_batch(self, batch):
        # Standard preprocessing
        return super().preprocess_batch(batch)
        
    def _do_train_epoch(self, loader):
        # Modified training loop with KD
        self.loss_names = ['box_loss', 'cls_loss', 'dfl_loss', 'kd_loss']  # Add KD loss
        # Rest of implementation...
```

If you'd like to contribute this feature, we'd welcome a PR implementing a proper distillation framework for YOLO11!

**Comment by Youho99** [comment]
@glenn-jocher 

An essential option would also be to be able to "separate" the images seen by the teacher and the student (noisy images, for example). But also, very importantly, to be able to give different image sizes to the teacher and the student (for example, 1280 for the teacher and 640 for the student).

This way, we can have a very good teacher, and see how the student can take advantage of the increased phantom resolution.

**Comment by Youho99** [comment]
I would like to work on this topic, but I think I would need to be accompanied by someone else who would be the leader of this PR.

**Comment by glenn-jocher** [comment]
@Youho99 Thank you for your interest in contributing! Your suggestion about different image sizes for teacher/student is valuable - this approach could indeed leverage the teacher's higher resolution insights while maintaining the student's efficiency.

The knowledge distillation implementation could include this feature through proper preprocessing in the training pipeline. Looking at the reference material on [Knowledge Distillation](https://www.ultralytics.com/glossary/knowledge-distillation), this approach aligns well with the core concept of transferring knowledge from larger to smaller models.

If you'd like to collaborate on this PR, perhaps we could create a discussion thread in the repository where interested contributors can coordinate efforts. I'd be happy to provide guidance while you and others implement the feature. The [customization documentation](https://docs.ultralytics.com/usage/engine/) shows how to properly extend the DetectionTrainer for this purpose.

Would you be interested in starting a discussion thread to coordinate this effort?

**Comment by Youho99** [comment]
@glenn-jocher yes you can create a discussion about it ! KD will benefit to everyone

**Comment by glenn-jocher** [comment]
I've created a discussion thread titled "Knowledge Distillation Implementation for YOLO11" where we can coordinate this effort: https://github.com/ultralytics/ultralytics/issues/19456

The knowledge distillation implementation should focus on several key aspects:
1. A flexible trainer architecture that supports different image sizes between teacher and student
2. Proper handling of soft targets from the teacher model
3. Customizable distillation loss weighting to balance standard supervision with teacher guidance

For those interested in contributing, please review the advanced customization documentation I shared earlier. The `DetectionTrainer` class provides extension points we can leverage without rebuilding the entire training pipeline.

Looking forward to seeing this collaborative effort come together. This will be a valuable addition to the Ultralytics ecosystem that can help users deploy more efficient models.

**Comment by cainiao123s** [comment]
> 我创建了一个名为“YOLO11 的知识提炼实现”的讨论帖，以便我们协调这项工作：[#19456](https://github.com/ultralytics/ultralytics/issues/19456)
> 
> 知识提炼的实施应该关注几个关键方面：
> 
> 1. 灵活的训练器架构，支持教师和学生之间不同的图像大小
> 2. 正确处理教师模型中的软目标
> 3. 可定制蒸馏损失权重，以平衡标准监督与教师指导
> 
> 有兴趣贡献代码的朋友，请查看我之前分享的高级定制文档。该`DetectionTrainer`文档提供了一些扩展点，我们可以利用这些扩展点，而无需重建整个训练流程。
> 
> 期待看到此次合作的成果。这将为 Ultralytics 生态系统带来宝贵的补充，帮助用户部署更高效的模型。

May I ask how I can stop backpropagation during the training process?

**Comment by Youho99** [comment]
@glenn-jocher maybe the hashtag id of the discussion is wrong ? It push me on a 3 month old closed issue

**Comment by cainiao123s** [comment]
> [@Youho99](https://github.com/Youho99) Thank you for your interest in contributing! Your suggestion about different image sizes for teacher/student is valuable - this approach could indeed leverage the teacher's higher resolution insights while maintaining the student's efficiency.
> 
> The knowledge distillation implementation could include this feature through proper preprocessing in the training pipeline. Looking at the reference material on [Knowledge Distillation](https://www.ultralytics.com/glossary/knowledge-distillation), this approach aligns well with the core concept of transferring knowledge from larger to smaller models.
> 
> If you'd like to collaborate on this PR, perhaps we could create a discussion thread in the repository where interested contributors can coordinate efforts. I'd be happy to provide guidance while you and others implement the feature. The [customization documentation](https://docs.ultralytics.com/usage/engine/) shows how to properly extend the DetectionTrainer for this purpose.
> 
> Would you be interested in starting a discussion thread to coordinate this effort?

And I would like to ask, does yolo Shuffle the dataset in each epoch during the training process

**Comment by glenn-jocher** [comment]
@Youho99 You're right, there seems to be an issue with that link. Let me create a new discussion thread specifically for knowledge distillation implementation and share the correct link shortly.

@cainiao123s To answer your questions:

1. To stop backpropagation for the teacher model during knowledge distillation, use the `.detach()` method on the teacher outputs as shown in the example above:
```python
teacher_preds = self.teacher_model(images).detach()
```
This prevents gradients from flowing through the teacher model.

2. Yes, YOLO does shuffle the dataset during training by default. The DataLoader in the Ultralytics framework has shuffle=True for training sets, which ensures different batch orderings across epochs for better generalization. You can verify this in the `get_dataloader` method of the BaseTrainer class.

I'll create and share a proper discussion thread for the knowledge distillation implementation shortly so everyone interested can collaborate effectively.

**Comment by cainiao123s** [comment]
Hi, I encountered some issues when doing knowledge distillation. For instance, when distilling feature maps, to align the feature map sizes of YOLOv8x and YOLOv8n, a trainable layer needs to be added to the student model (to change the number of channels). Without modifying the YOLOv8n model, how can I achieve this?

**Comment by cainiao123s** [comment]
Hi, I encountered some issues when doing knowledge distillation. I want to add a trainable parameter in the class v8DetectionLoss: function. How can I achieve this?

**Comment by glenn-jocher** [comment]
@cainiao123s To add trainable parameters to the loss function for knowledge distillation, you can subclass the loss class and register parameters using `nn.Parameter`. Here's how:

```python
import torch.nn as nn

class KDDetectionLoss(v8DetectionLoss):
    def __init__(self, model, tau=4.0):
        super().__init__(model)
        # Add trainable parameters
        self.alpha = nn.Parameter(torch.tensor(0.5))  # Weight for KD loss
        self.tau = tau  # Temperature for softmax
        
    def __call__(self, preds, batch, teacher_preds=None):
        # Standard detection loss
        loss = super().__call__(preds, batch)
        
        # Add KD loss if teacher predictions provided
        if teacher_preds is not None:
            kd_loss = self.compute_kd_loss(preds, teacher_preds)
            loss += self.alpha * kd_loss
            
        return loss
```

The `nn.Parameter` wrapper makes `alpha` trainable and automatically registers it with the model's parameters. You can then use this custom loss in your `DetectionTrainer` subclass by overriding the `init_criterion` method.
